tw-tasks-core/src/main/java/com/transferwise/tasks/health/ClusterWideTasksStateMonitor.java

```diff
@@ -84,33 +89,49 @@ public void init() {
                   taskHandleHolder.getValue().stop();
                   taskHandleHolder.getValue().waitUntilStopped(Duration.ofMinutes(1));
                 }
-                resetState();
+                resetState(false);
                 log.info("Monitoring of tasks state stopped.");
               });
         });
   }
 
-  protected void resetState() {
-    if (registeredMetricHandles != null) {
-      for (Object metricHandle : registeredMetricHandles) {
-        meterHelper.unregisterMetric(metricHandle);
+  protected void resetState(boolean forInit) {
```

Comment:
Maybe we should have 2 methods and avoid lazy initialisation in `checkX` methods. For example

```java
void initialize() {
  stateLock.lock();
  try {
    // assign all state variables to new values
    erroneousTasksCount = new AtomicInteger(0);

    // register all gauges
    registeredMetricHandles.add(
        meterHelper.registerGauge(METRIC_PREFIX + "health.tasksInErrorCount", erroneousTasksCount::get)
    );

    initialized = true;
  } finally {
    stateLock.unlock();
  }
}

void terminate() {
  stateLock.lock();
  try {
    initialized = false;

    // unregister metrics
    
    // reset variables
  } finally {
    stateLock.unlock();
  }
}
```

Reply:
There is another idea behind the lazy-init we are doing - we want to avoid blips in metrics.

If I would register a gauge here with value 0, it could cause a blip, when we actually had some tasks in error.